# PHACTransform
CS-515 Deep Learning class project

## Abstract

Mutations are pivotal in biology, driving genetic diversity and influencing various biological processes. Predicting the effects of mutations accurately remains challenging due to genetic complexity and vast genomic data. Deep Mutational Scanning (DMS) offers insights into protein function and genetic diversity but is resource-intensive. Computational tools like EVE (Evolutionary Model of Variant Effect) and Rep2Mut-V2 leverage sequence alignments and deep learning to predict mutation effects efficiently. However, these methods often lack precision in capturing biological complexities.
We introduce PHACTransform, a novel approach integrating multiple sequence alignments (MSA), evolutionary data, and experimental DMS scores into deep learning models for precise variant effect prediction. By incorporating PHACT-based measures and DMS scores, our approach aims to enhance mutation effect predictions beyond conventional methods. Our research contributes to advancing genetic variation understanding, impacting drug discovery, personalized medicine, and evolutionary biology.
